### PR TITLE
Merge pytorch, tf2, and carla mot images

### PR DESCRIPTION
**Pull request by christopherwoodall**
- [x] Consolidates Armory's `pytorch`, `tf2`, and `carla` images
- [x] Remove old images
- [x] Add new image to CI
- [x] Update hash map
- [x] Update `docker/build.py`
- [ ] Update dependencies in `pyproject.toml` and `environment.yaml`
- [x] Bump minimum Python version to 3.8 
- [x] Update documentation
- [ ] Fix `--interactive` displaying `i have no name` as the username (1)
- [ ] Expose ports  in Dockerfile (1)
- [ ] List volumes in Dockerfile (1)

1. These changes should eventually be moved into the base image, but that exceeds the scope of this PR.

### CODE REVIEW

**Comment by davidslater** [comment]
If we are going to move in this direction, let's also remove `tf2` and `pytorch-deepspeech`, and just move directly to one container.

**Comment by davidslater** [comment]
`TF2` can be mapped to `PYTORCH`. Actually, I would probably just map them all to `armory` or some generic container name.

the deep speech container will require some additional work:
a) removal of `armory/baseline_models/pytorch/deep_speech.py`
b) pinning of `eval1-4/asr_librispeech/`, `eval5/asr_librispeech/` to 0.16.1 version docker container (and adding deprecation)
c) expanding `eval6/asr_librispeech` to include targeted and targeted entailment attacks
d) various updates of docs

**Comment by christopherwoodall** [comment]
Added fix for #1787

**Comment by christopherwoodall** [comment]
I agree with the above, but I think we should scope that work out into a different PR.

**Comment by davidslater** [comment]
Let's have `pytorch-deepspeech` be a different PR, but let's put `tf2` in this one, mapping it to PyTorch like you have done here with `carla-mot`. The current `tf2` and `pytorch` docker images are effectively identical.

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [project]
-requires-python = ">=3.7"
+requires-python = ">=3.8"
```
Comment (mwartell) [inline pyproject.toml:2]: pure fyi: the reason we deprecated 3.7 support was because we used `ubuntu-latest` in CI and when Github Actions bumped latest == 22.04 we lost python 3.7 there. While we could have patched it in from the `deadsnakes` PPA we thought that silly.

And, since we shouldn't claim support for that which we don't (or can't) test, then it is proper for us to declare 3.8+

File: pyproject.toml [hunk 2]
```diff
@@ -115,8 +106,34 @@ deepspeech = [
...
+
+all = [
+    "armory-testbed[developer]",
+    "armory-testbed[engine]",
+    "armory-testbed[math]",
+    "armory-testbed[jupyter]",
+    "armory-testbed[datasets]",
+    "armory-testbed[datasets-builder]",
+    "armory-testbed[pytorch]",
+    # "armory-testbed[deepspeech]",
```
Comment (mwartell) [inline pyproject.toml:132]: dead code; should get a comment or be deleted

File: pyproject.toml (hunk 1 above)
Comment (christopherwoodall) [reply pyproject.toml:2]: I'd like to add that support for 3.7 ends in 6 months as well.

File: armory/docker/images.py [hunk 3]
```diff
@@ -15,22 +15,18 @@
...
-    PYTORCH,
+    ARMORY_IMAGE_NAME,
     PYTORCH_DEEPSPEECH,
-    TF2,
-    CARLA_MOT,
 )
 REPOSITORIES = tuple(x.split(":")[0] for x in ALL)
 IMAGE_MAP = {
-    "pytorch": PYTORCH,
+    "pytorch": ARMORY_IMAGE_NAME,
```
Comment (davidslater) [inline armory/docker/images.py:26]: Can you add a key-value pair `"armory": ARMORY_IMAGE_NAME,`?

This dict is used in `__main__` and in image lookups, and this would enable using `armory` on the command line or the scenario configs.

File: pyproject.toml [hunk 4]
```diff
@@ -55,50 +54,36 @@ developer =[
...
+    "tensorboardx == 2.5.1",
 ]
 
 pytorch = [
     "armory-testbed[engine,datasets,math]",
     "torch",
-    "torchvision",
-    "torchaudio",
+    "torchaudio == 0.12.1",
+    "torchvision == 0.13.1",
```
Comment (davidslater) [inline pyproject.toml:79]: Why are these specified? If they are specified, `torch` version should be specified as well.

File: environment.yml [hunk 5]
```diff
@@ -1,24 +1,42 @@
...
+
+  # ByteTrack (see below)
+  # TODO: Should this be added to pyproject.toml; e.g. [byte-track]?
+  - pytorch=1.12   # Required by ByteTrack, therefore must be installed pre-pip
+  - Cython         # ByteTrack
+  - lap            # ByteTrack
+
+  - pip:
+    # Hydra Lightning Configs
+    - 'git+https://github.com/romesco/hydra-lightning.git#subdirectory=hydra-configs-pytorch-lightning'
```
Comment (davidslater) [inline environment.yml:34]: No, this is part of pytorch-deepspeech.

File: environment.yml [hunk 6]
```diff
@@ -1,24 +1,42 @@
...
-# yamllint disable rule:line-length
 name: armory-core
+
 channels:
-  - pytorch
-  - conda-forge
   - defaults
+  - conda-forge
+  - anaconda
+  - pytorch
```
Comment (davidslater) [inline environment.yml:8]: Why is the ordering changed?

File: pyproject.toml [hunk 7]
```diff
@@ -55,50 +54,36 @@ developer =[
...
+    "boto3 == 1.24.89",      # Needed for armory.data.utils
+    "botocore == 1.27.89" ,  # Needed for armory.data.utils
+    "ffmpeg-python",         # Needed for armory.utils.export
+    "pydub",               # this is in ART's extra-requires
+    "tidecv",              # Needed for TIDE metrics
+    # Both `opencv-python` and `opencv-python-headless` must specify
+    # the same version.
+    "opencv-python == 4.5.5.62",          # Needed for CARLA baseline scenario
+    "opencv-python-headless == 4.5.5.62", # Needed for CARLA baseline scenario
+    "tensorboardx == 2.5.1",
```
Comment (davidslater) [inline pyproject.toml:72]: These are very specific version numbers. Why?

File: pyproject.toml [hunk 8]
```diff
@@ -113,10 +98,43 @@ deepspeech = [
...
+    "armory-testbed[tensorflow]",
+    "tensorflow-datasets >= 4.6.0",
+    "protobuf",
+]
+
+datasets-builder = [
+    "armory-testbed[datasets,engine]",
+    "apache-beam >= 2.22.0",
+]
+
```
Comment (davidslater) [inline pyproject.toml:119]: Is there a reason these all got reordered? Hard to follow changes.

File: environment.yml [hunk 9]
```diff
@@ -1,24 +1,42 @@
 ---
-# yamllint disable rule:line-length
 name: armory-core
```
Comment (davidslater) [inline environment.yml:2]: Changing the `environment.yml` will impact the base image. Not something we want to do with this mod.

**Review by davidslater — CHANGES_REQUESTED** [review]
I have a lot of questions - probably better to discuss in person.

File: pyproject.toml (hunk 7 above)
Comment (christopherwoodall) [reply pyproject.toml:72]: In order to speed up pip's resolver as well as to ensure compatibility with all the dependencies.

File: environment.yml (hunk 9 above)
Comment (christopherwoodall) [reply environment.yml:2]: The changes do effect the base image, but all images continue to build as normal.

File: docker/Dockerfile-pytorch [hunk 10]
```diff
@@ -21,7 +21,9 @@ RUN pip install --no-cache-dir \
         git+https://github.com/ifzhang/ByteTrack.git \
         thop \
         lap \
-        Cython && \
+        Cython \
+        # BURST dependency
+        tabulate && \
```
[the remark this answers is not in the document]
Comment (christopherwoodall) [reply docker/Dockerfile-pytorch]: There is a dependency warning if not installed. My guess is that it is required by another package.

File: pyproject.toml (hunk 8 above)
Comment (christopherwoodall) [reply pyproject.toml]: This change moves the frameworks to the top so that it is easier to find their dependencies.

**Review by davidslater — CHANGES_REQUESTED** [review]
Please make this into 2 PRs:
1) A PR that does the basic changes to workflows, docker images, __main__, build and docs, but doesn't add any new features to `Dockerfile-armory` (e.g., docker user or port exposure) or make any mods to `environment.yml` or `pyproject.toml`.
2) A second PR (that can branch from the first one) that adds the new features, environmental changes, etc.?

**Comment by christopherwoodall** [comment]
Split PR. See PR #1807

File: armory/docker/images.py [hunk 11]
```diff
@@ -10,27 +10,19 @@
...
-ALL = (
-    PYTORCH,
-    PYTORCH_DEEPSPEECH,
-    TF2,
-    CARLA_MOT,
-)
-REPOSITORIES = tuple(x.split(":")[0] for x in ALL)
+TAG = version.to_docker_tag(armory.__version__)
+ARMORY_IMAGE_NAME = f"twosixarmory/armory:{TAG}"
+DEEPSPEECH_IMAGE_NAME = f"twosixarmory/deepspeech:{TAG}"
```
Comment (davidslater) [inline armory/docker/images.py:18]: Should be f"twosixarmory/pytorch-deepspeech:{TAG}"

File: environment.yml [hunk 12]
```diff
@@ -21,4 +21,4 @@ dependencies:
   - numba  # ART requires numba >= 0.53.1
   - pandas
   - protobuf
-prefix: /opt/conda
+prefix: /opt/conda
```
Comment (davidslater) [inline environment.yml:24]: Add back in EOL

File: pyproject.toml [hunk 13]
```diff
@@ -55,50 +61,35 @@ developer =[
...
+    "Pillow",              # Data dependencies
+    "boto3",               # Needed for armory.data.utils
+    "botocore" ,           # Needed for armory.data.utils
+    "ffmpeg-python",       # Needed for armory.utils.export
+    "pydub",               # this is in ART's extra-requires
+    "tidecv",              # Needed for TIDE metrics
+    # Both `opencv-python` and `opencv-python-headless` must specify
+    # the same version.
+    "opencv-python == 4.5.5.62",          # Needed for CARLA baseline scenario
+    "opencv-python-headless == 4.5.5.62", # Needed for CARLA baseline scenario
```
Comment (davidslater) [inline pyproject.toml:78]: Remind me why `opencv-python-headless` is needed

Comment (davidslater) [inline pyproject.toml:78]: Can we remove this pinning for this part and have it for part II instead?

**Review by davidslater — APPROVED** [review]
Two minor edits - but feel free to merge after fixing those.

Prior to merging, please build the armory container and push to dockerhub as `twosixarmory/armory:0.16.1`.
